include/sparrow/dynamic_bitset.hpp

```diff
@@ -0,0 +1,338 @@
+
+        void resize(size_type n, value_type = false);
+
+        size_type compute_block_count(size_type bits_count) const noexcept;
+
+    private:
+
+        size_type count_non_null() const noexcept;
+        size_type block_index(size_type pos) const noexcept;
+        size_type bit_index(size_type pos) const noexcept;
+        block_type bit_mask(size_type pos) const noexcept;
+        size_type count_extra_bits() const noexcept;
+        void zero_unused_bits();
+
+        static constexpr std::size_t s_bits_per_block = sizeof(block_type) * CHAR_BIT;
+
+        storage_type m_buffer;
+        size_type m_size;
+        size_type m_null_count;
+    };
+
+    template <class T>
+    class dynamic_bitset : public dynamic_bitset_base<buffer<T>>
+    {
+    public:
+
+        using base_type = dynamic_bitset_base<buffer<T>>;
+        using storage_type = typename base_type::storage_type;
+        using block_type = typename base_type::block_type;
+        using value_type = typename base_type::size_type;
+        using size_type = typename base_type::size_type;
+
+        dynamic_bitset();
+        explicit dynamic_bitset(size_type n, value_type v = false);
+        dynamic_bitset(block_type* p, size_type n);
+        dynamic_bitset(block_type* p, size_type n, size_type null_count);
+
+        ~dynamic_bitset() = default;
+        dynamic_bitset(const dynamic_bitset&) = default;
+        dynamic_bitset(dynamic_bitset&&) = default;
+
+        dynamic_bitset& operator=(const dynamic_bitset&) = default;
+        dynamic_bitset& operator=(dynamic_bitset&&) = default;
+
+        using base_type::resize;
+    };
+
+    template <class T>
+    class dynamic_bitset_view : public dynamic_bitset_base<buffer_view<T>>
+    {
+    public:
+
+        using base_type = dynamic_bitset_base<buffer_view<T>>;
+        using storage_type = typename base_type::storage_type;
+        using block_type = typename base_type::block_type;
+        using size_type = typename base_type::size_type;
+
+        dynamic_bitset_view(block_type* p, size_type n);
+        dynamic_bitset_view(block_type* p, size_type n, size_type null_count);
+        ~dynamic_bitset_view() = default;
+
+        dynamic_bitset_view(const dynamic_bitset_view&) = default;
+        dynamic_bitset_view(dynamic_bitset_view&&) = default;
+
+        dynamic_bitset_view& operator=(const dynamic_bitset_view&) = default;
+        dynamic_bitset_view& operator=(dynamic_bitset_view&&) = default;
+    };
+
+    /**************************************
+     * dynamic_bitset_base implementation *
+     **************************************/
+
+    template <class B>
+    auto dynamic_bitset_base<B>::size() const noexcept -> size_type 
+    {
+        return m_size;
+    }
+
+    template <class B>
+    auto dynamic_bitset_base<B>::null_count() const noexcept -> size_type
+    {
+        return m_null_count;
+    }
+
+    template <class B>
+    bool dynamic_bitset_base<B>::test(size_type pos) const
+    {
```

Comment:
Consider adding assertion on all functions that could fail because of incorrect input.